### DISCUSSION TITLE
CREPA and other stuff

### DISCUSSION THREAD

**Discussion by dorpxam** [AkaneTendo25/musubi-tuner#94] [General]
I'm always in a understanding phase to see how this kind of enhancement works. It's difficult to really understand what is the impact of CREPA, at least in a scientific sense. So everything I'm going to say here is based on feelings and offhand observations.

## Latest test

One latest test using both CREPA/Prior Divergence, strangely, shows a nearly flat loss/video. It's important to note that this is a retraining session, and I'm starting this training from a previous checkpoint. Yet, against all expectations, the model does not saturate and is still learning. The LoRA shows no signs of overfitting and the quality remains excellent.

https://github.com/user-attachments/assets/5f3f73fe-af66-4b80-9241-ea1d0f94ab33

The optimizer settings for this one is:
```python
--optimizer_type AdamW
--optimizer_type schedulefree.AdamWScheduleFree
--optimizer_args "weight_decay=0.001" "betas=(0.9,0.95)"
```
And the parameters of CREPA/prior_div:
```python
--prior_divergence
--prior_divergence_args multiplier=0.1
--crepa
--crepa_args mode=backbone student_block_idx=16 teacher_block_idx=32 lambda_crepa=0.5 tau=1.0 num_neighbors=2 schedule=cosine warmup_steps=300 normalize=true
```
It would appear that `weight_decay` helps to avoid overfitting.

## CREPA / Prodigy

I have experimented the couple CREPA/Prodigy and that works too. Here the optimizer settings:

```python
--learning_rate 1 ^
--optimizer_type prodigyopt.Prodigy
--optimizer_args "weight_decay=0.01" "decouple=True" "safeguard_warmup=False" "use_bias_correction=True"
--lr_scheduler_type "CosineAnnealingLR"
--lr_scheduler_args "T_max=3072"
```

## Loss Reading

The biggest problem with CREPA/Prior Divergence is that the negative aspect of the values ​​makes interpreting the losses almost unfair. Here's the curve from the last test with a flat loss/video, and yet the improvement from the previous training is very clear.

<img width="2500" height="1197" alt="Image" src="https://github.com/user-attachments/assets/26e7611d-1ae5-420f-a62c-cc86e3d11a59" />

I'm trying to understand how CREPA works in order to improve its use. For example, the parameter `max_steps=0` makes CREPA active throughout the entire training. Is this actually useful? What happens if I limit CREPA to half or even a third of the training steps?

## Differential Output Preservation (DOP)

As you can see in the video above, the character is duplicated and appears at the end. This is where DoP makes sense and could prevent this kind of problem. However, using regularization makes training incredibly slow (> 20 sec/it).

My question might be silly and naive, but wouldn't it be possible, for example, to have this kind of scenario:

- Phase 1: Training with CREPA (and potentially Prior Divergence, although I'm not sure if the impact is significant)
- Phase 2: Training (retraining) the LoRa with a limited number of steps but with DoP enabled.

Could this work? The idea is to get the benefits of DoP without having to apply it to the entire training process.

It might seem silly, but if it works, it could be really interesting for basic GPUs like mine.

## Next Test
The next training session will be with Prodigy and CREPA. This combination seems promising.

I'm going to try the CREPA parameters inspired by [this one](https://huggingface.co/kabachuha/ltx2-hydraulic-press):

```python
--crepa
--crepa_args mode=backbone student_block_idx=8 teacher_block_idx=42 lambda_crepa=0.02 tau=1.0 num_neighbors=2 schedule=cosine warmup_steps=100 max_steps=0 normalize=true
```

Since the parameter set differs from that of SimpleTuner, this is a purely blind test. 

If you have any advice for me, I'm all ears. I think these options are very useful but require a level of expertise that's a bit beyond my capabilities.

Once again, thank you for everything.

**Comment by dorpxam** [comment]
CREPA in low lambda (aka `0.02` here), same parameters than previous 'Next Test' message except attempt to put `num_neighbors=4` and `max_steps` to half the training steps (`1250`).

Seem more coherent loss graph (positive average), not tested yet (always in training).

<img width="2485" height="1221" alt="Image" src="https://github.com/user-attachments/assets/0f683abf-a8d5-4881-be88-e045b82611ac" />

**Comment by AkaneTendo25** [comment]
I think that a flat average loss curve is expected when using CREPA, since there are several loss components (base diffusion loss, CREPA loss, prior regularization loss), some of them are negative, so it's better to estimate their progress by separate loss graphs or raw visual output quality. 

I cannot say much about combining CREPA with other regularization techniques, in theory, prior divergence and CREPA should be complementary. Constraining CREPA + prior to a first stage with a consequent DOP stage seems like a good idea.

Most important CREPA params are `lambda_crepa`, which is basically the CREPA loss weight, and too high value may make LoRA effects weak (temporal coherence factor outweighs the main training objective), and too low renders CREPA useless. Another valuable one is `num_neighbors`, which sets how many frames are aligned. Increasing it enforces learning long-clip consistency; low values are more suited to short-term consistency. `schedule` and `warmup_steps` are important if CREPA interferes with early training too much and prevents actual LoRA training. Remember that CREPA is a regularizer technique, so it cannot just magically make LoRA training stable.

About the use case of constant CREPA (enforced by `max_steps=0`), it might be probably helpful for simplicity, for short training runs, or when you want to enforce temporal coherence on an already trained LoRA - that is what I can come up with. Probably not very useful for mainstream cases, so using a cosine decay schedule is preferable. (Btw, you're passing `schedule=cosine`, but it is ignored because `max_steps=0`.)

I did not experiment much with these methods (most likely, I wouldn't even implement them unless you told me about them), so I can't really be helpful here yet :)

**Comment by dorpxam** [comment]
OH ! Cool. Some very useful informations. Yes, I understand that is not magically, but anyway I'm sure there's a good and useful 'add-on'. It's difficult for now to have a clear view of the impact of the settings, I need more investigation.

Thanks for the reply.
